generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -189,16 +203,21 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
     if (!shadingLanguageVersion.supportedBitwiseOperations()) {
       return Optional.empty();
     }
-    final BinOp operator = generator.nextBoolean() ? BinOp.SHL : BinOp.SHR;
-    // The maximum safe value to shift zero without potentially causing undefined behavior is
-    // between zero and eight bits (as nine bits is the minimum precision of an integer in GLSL).
-    // If we're shifting one, we want to be able to shift backwards without losing information, and
-    // so our maximum shift value is seven bits.
-    final int bound = isZero ? 10 : 9;
-    final int maxValue = generator.nextInt(bound);
+    // The minimum precision for a lowp integer in GLSL is 9 bits (with one reserved for a sign
+    // bit if the integer is signed) - we don't have to worry about losing information if we're
+    // shifting zero, but shifting more than 8 bits may result in undefined behavior.
+    final int maxShiftForZero = 9;
+    // While we still have a hard maximum bits to shift of 8 bits, if we're shifting one, then
+    // we can potentially lose information because GLSL bit shifting is not circular. To remedy
+    // that, we make sure not to shift our 1 bit out of the integer by limiting our maximum shift
+    // to 7 bits.
```

Comment:
Then I would rename this one to "minNonSignBitsForLowpInt", and initialize it to "minBitsForLowpInt - 1" (rather than explicitly to 7).

Reply:
Done.